tests/org.eclipse.ui.tests/Eclipse UI Tests/org/eclipse/ui/tests/preferences/ViewerItemsLimitTest.java

```diff
@@ -0,0 +1,470 @@
+		assertEquals("3 project error has occured", firstItem.getItems()[4].getText());
+
+		setNewViewerLimit(DEFAULT_VIEW_LIMIT);
+
+		// job may take some time to create marker.
+		processBackgroundUpdates(100);
+
+		firstItem = getFirstItem(commonViewer);
+		assertEquals("all the items must be visible with limit more than input", numberOfMarkers,
+				firstItem.getItems().length);
+	}
+
+	private TreeItem getFirstItem(MarkersTreeViewer commonViewer) {
+		TreeItem[] items = commonViewer.getTree().getItems();
+		return items.length > 0 ? items[0] : null;
+	}
+
+	/**
+	 * Create some projects and some search data. Create some query and execute the
+	 * search. Check limited search results. Add new search data onto workspace and
+	 * refresh search to see if search results updated properly.
+	 *
+	 * @throws CoreException
+	 */
+	@Test
+	public void testLimitedSearchResult() throws CoreException {
+		IWorkspace workspace = ResourcesPlugin.getWorkspace();
+		IWorkspaceRoot workspaceRoot = workspace.getRoot();
+		setNewViewerLimit(VIEW_LIMIT_3);
+		int numOfProj = 9;
+		for (int i = 1; i <= numOfProj; i++) {
+			createProjectWithFile(workspaceRoot, i);
+		}
+		processBackgroundUpdates(100);
+
+		SearchView searchView = (SearchView) activePage.showView(SEARCH_VIEW_ID);
+		FileTextSearchScope scope = FileTextSearchScope.newWorkspaceScope(new String[] { "*" }, false);
+		FileSearchQuery searchQuery = new FileSearchQuery("some", false, false, false, false, scope);
+		NewSearchUI.runQueryInBackground(searchQuery);
+
+		// job may take some time to create marker.
+		processBackgroundUpdates(1000);
+
+		FileSearchPage searchPage = (FileSearchPage) searchView.getActivePage();
+		assertNotNull("Search page should be shown", searchPage);
+		TreeViewer viewer = (TreeViewer) searchPage.getViewer();
+		Tree tree = viewer.getTree();
+		processEventsUntil(() -> tree.getItems().length > VIEW_LIMIT_3, 30_000);
+
+		TreeItem[] items = tree.getItems();
+		assertLimitedItems(VIEW_LIMIT_3, numOfProj, items);
+		assertFirstSearchResultExpanded(items, "jp1", VIEW_LIMIT_3);
+
+		// this is equal to refresh of search results.
+		createProjectWithFile(workspaceRoot, 0);
+		processBackgroundUpdates(1000);
+
+		numOfProj++;
+		NewSearchUI.runQueryInBackground(searchQuery);
+		processBackgroundUpdates(1000);
+
+		processEventsUntil(() -> tree.getItems().length > VIEW_LIMIT_3, 30_000);
+		items = tree.getItems();
+		assertLimitedItems(VIEW_LIMIT_3, numOfProj, items);
+		assertFirstSearchResultExpanded(items, "jp0", VIEW_LIMIT_3);
+
+		// this is equal to refresh of search results.
+		createProjectWithFile(workspaceRoot, 50);
+		processBackgroundUpdates(1000);
+
+		numOfProj++;
+		NewSearchUI.runQueryInBackground(searchQuery);
+		processBackgroundUpdates(1000);
+
+		processEventsUntil(() -> tree.getItems().length > VIEW_LIMIT_3, 30_000);
+		items = tree.getItems();
+		assertLimitedItems(VIEW_LIMIT_3, numOfProj, items);
```

Comment:
Interestingly, one random fail on Windows here: we see more items as expected???

```
**items length mist be limit plus one expected:<4> but was:<6>
junit.framework.AssertionFailedError: items length mist be limit plus one expected:<4> but was:<6>
	at junit.framework.Assert.fail(Assert.java:57)
	at junit.framework.Assert.failNotEquals(Assert.java:329)
	at junit.framework.Assert.assertEquals(Assert.java:78)
	at junit.framework.Assert.assertEquals(Assert.java:234)
	at junit.framework.TestCase.assertEquals(TestCase.java:377)
	at org.eclipse.ui.tests.preferences.ViewerItemsLimitTest.assertLimitedItems(ViewerItemsLimitTest.java:237)
	at org.eclipse.ui.tests.preferences.ViewerItemsLimitTest.testLimitedSearchResult(ViewerItemsLimitTest.java:412)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
```

Reply:
Failed again, and again **on Windows only**. On Linux I can't see anything wrong...
I wonder if the Windows Tree/Table shows some different behavior that breaks the test.
@raghucssit : please try to debug this test **on Windows**.
```
items length mist be limit plus one expected:<4> but was:<7>
junit.framework.AssertionFailedError: items length mist be limit plus one expected:<4> but was:<7>
	at junit.framework.Assert.fail(Assert.java:57)
	at junit.framework.Assert.failNotEquals(Assert.java:329)
	at junit.framework.Assert.assertEquals(Assert.java:78)
	at junit.framework.Assert.assertEquals(Assert.java:234)
	at junit.framework.TestCase.assertEquals(TestCase.java:377)
	at org.eclipse.ui.tests.preferences.ViewerItemsLimitTest.assertLimitedItems(ViewerItemsLimitTest.java:237)
	at org.eclipse.ui.tests.preferences.ViewerItemsLimitTest.testLimitedSearchResult(ViewerItemsLimitTest.java:412)
```

Reply:
Disabled search test and reverted incremental display limit for search view.

Reply:
Can you please elaborate why the search result view can handle so many items but the package explorer cannot?